### PR TITLE
Added typeshape mapper

### PR DESCRIPTION
**Pull request by vshapenko**
This PR adds typeshape mapper support for LiteDb.Fsharp. Currently it passes 76 of 78 tests

### CODE REVIEW

**Comment by Zaid-Ajaj** [comment]
Awesome! can you add your mapper to the tests to see which ones are failing? There is this function:
```fs
let useDatabase (f: LiteDatabase -> unit) = 
    let mapper = FSharpBsonMapper()
    use memoryStream = new MemoryStream()
    use db = new LiteDatabase(memoryStream, mapper)
    f db
```
create another one `useTypeShapeDb`:
```fs
let useTypeShapeDb(f: LiteDatabase -> unit) = 
    let mapper = Experiemental.TypeShapeMapper()
    use memoryStream = new MemoryStream()
    use db = new LiteDatabase(memoryStream, mapper)
    f db
```
Then parameterize the tests such that they take `(useDatabase : LiteDatabase -> unit)` parameter, what do you think?

File: LiteDB.FSharp.Tests/Tests.InheritedType.fs [hunk 1]
```diff
@@ -100,7 +101,7 @@ type Item2OfRecord =
 
 
 let useDatabase (f: LiteRepository -> unit) = 
-    let mapper = FSharpBsonMapper()
+    let mapper = new TypeShapeMapper()
```
Comment (Zaid-Ajaj) [inline LiteDB.FSharp.Tests/Tests.InheritedType.fs]: The idea is not to replace but to let the same tests run for both mappers

**Comment by vshapenko** [comment]
@Zaid-Ajaj  Based on tests, we got 3x performance gain

File: LiteDB.FSharp.Tests/Tests.InheritedType.fs (hunk 1 above)
Comment (vshapenko) [reply LiteDB.FSharp.Tests/Tests.InheritedType.fs]: Refactoring made, my mistake, too early push

**Comment by Zaid-Ajaj** [comment]
Wohoo Nice! 🚀 it looks like these tests are failing:
 - [Full custom search works by BsonValue deserialization](https://github.com/Zaid-Ajaj/LiteDB.FSharp/blob/master/LiteDB.FSharp.Tests/Tests.LiteDatabase.fs#L565)
 - [Full custom search works by using expressions](https://github.com/Zaid-Ajaj/LiteDB.FSharp/blob/master/LiteDB.FSharp.Tests/Tests.LiteDatabase.fs#L590)

I guess that is because `Bson.deserialize` and `Bson.serialize` are still mapper-dependent, can you parameterize these as well such that take and use a mapper as input? 

Lastly, it would be great to add a little section to the README regarding how to use the type-shape mapper 🙏

**Comment by vshapenko** [comment]
> Wohoo Nice! 🚀 it looks like these tests are failing:
> 
> * [Full custom search works by BsonValue deserialization](https://github.com/Zaid-Ajaj/LiteDB.FSharp/blob/master/LiteDB.FSharp.Tests/Tests.LiteDatabase.fs#L565)
> * [Full custom search works by using expressions](https://github.com/Zaid-Ajaj/LiteDB.FSharp/blob/master/LiteDB.FSharp.Tests/Tests.LiteDatabase.fs#L590)
> 
> I guess that is because `Bson.deserialize` and `Bson.serialize` are still mapper-dependent, can you parameterize these as well such that take and use a mapper as input?
> 
> Lastly, it would be great to add a little section to the README regarding how to use the type-shape mapper 🙏

Bson.serialize and deserialize do not depend on mapper

**Comment by Zaid-Ajaj** [comment]
I see, I though we could make Bson.fs mapper-dependent but that is OK, I guess we can merge this, I will put the Bson specific tests in their own testList that only takes FSharpBsonMapper

**Comment by Zaid-Ajaj** [comment]
I will add a README section as well, will be able to work on this later this evening

**Comment by Zaid-Ajaj** [comment]
Published as of v2.12.0 :rocket:

**Pull request merged**